crates/sui-framework/src/natives/dynamic_field.rs

```diff
@@ -0,0 +1,267 @@
+
+use crate::{
+    legacy_emit_cost,
+    natives::{
+        get_object_id,
+        object_runtime::{object_store::ObjectResult, ObjectRuntime},
+    },
+};
+use digest::Digest;
+use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_core_types::{
+    account_address::AccountAddress,
+    language_storage::{StructTag, TypeTag},
+    value::MoveTypeLayout,
+    vm_status::StatusCode,
+};
+use move_vm_runtime::native_functions::NativeContext;
+use move_vm_types::{
+    loaded_data::runtime_types::Type, natives::function::NativeResult, pop_arg, values::Value,
+};
+use sha3::Sha3_256;
+use smallvec::smallvec;
+use std::collections::VecDeque;
+use sui_types::base_types::{ObjectID, SuiAddress};
+
+const E_KEY_ALREADY_EXISTS: u64 = 0;
+const E_KEY_DOES_NOT_EXIST: u64 = 1;
+const E_FIELD_TYPE_MISMATCH: u64 = 2;
+const E_BCS_SERIALIZATION_FAILURE: u64 = 3;
+
+macro_rules! get_or_fetch_object {
+    ($context:ident, $ty_args:ident, $parent:ident, $child_id:ident) => {{
+        let child_ty = $ty_args.pop().unwrap();
+        assert!($ty_args.is_empty());
+        let (layout, tag) = match get_tag_and_layout($context, &child_ty)? {
+            Some(res) => res,
+            None => {
+                return Ok(NativeResult::err(
+                    legacy_emit_cost(),
+                    E_BCS_SERIALIZATION_FAILURE,
+                ))
+            }
+        };
+        let object_runtime: &mut ObjectRuntime = $context.extensions_mut().get_mut();
+        object_runtime.get_or_fetch_child_object($parent, $child_id, &child_ty, layout, tag)?
+    }};
+}
+
+// native fun hash_type_and_key<K: copy + drop + store>(parent: address, k: K): address;
+pub fn hash_type_and_key(
+    context: &mut NativeContext,
+    mut ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    assert!(ty_args.len() == 1);
+    assert!(args.len() == 2);
+    let k_ty = ty_args.pop().unwrap();
+    let k: Value = args.pop_back().unwrap();
+    let parent: SuiAddress = pop_arg!(args, AccountAddress).into();
+    let k_tag = context.type_to_type_tag(&k_ty)?;
+    // build bytes
+    let k_tag_bytes = match bcs::to_bytes(&k_tag) {
+        Ok(bytes) => bytes,
+        Err(_) => {
+            return Ok(NativeResult::err(
+                legacy_emit_cost(),
+                E_BCS_SERIALIZATION_FAILURE,
+            ));
+        }
+    };
+    let k_layout = match context.type_to_type_layout(&k_ty) {
+        Ok(Some(layout)) => layout,
+        _ => {
+            return Ok(NativeResult::err(
+                legacy_emit_cost(),
+                E_BCS_SERIALIZATION_FAILURE,
+            ))
+        }
+    };
+    let k_bytes = match k.simple_serialize(&k_layout) {
+        Some(bytes) => bytes,
+        None => {
+            return Ok(NativeResult::err(
+                legacy_emit_cost(),
+                E_BCS_SERIALIZATION_FAILURE,
+            ))
+        }
+    };
+    // hash(parent || k || K)
```

Comment:
Probably also want a domain separator here--will let @kchalkias weigh in.

Reply:
Actually: I think the cleanest thing is probably to add `ObjectID::new_for_dynamic_field(self, k: &TypeTag, value: Vec<u8>)` + put it right next to the other code for deriving an object ID via hashing?

Reply:
As discussed w @tnowacki the domain separation epic is under finalization. Please add a TODO comment and assign it to myself, and we'll soon update all hashes and sigs with intents, which will effectively be a breaking change in other places anyway. Atm we're good without requiring to improvise here. Imagine the intent for hashes as being a prefix flag (a couple of bytes) for each hash operation, thus for any new functionality we'll pick a new flag and thus we won't expect cross-domain collisions.